src/mat/impls/inv/matinv.c

```diff
@@ -402,16 +370,54 @@ static PetscErrorCode MatInvReset_Inv(Mat imat)
   PetscFunctionReturnI(0);
 }
 
+//TODO propose to PETSc function comparing id and state at the same time
+//TODO we can take R from inv-A
+static PetscErrorCode MatInvImprintNullSpace_Inv(Mat imat, Mat R)
+{
+  Mat_Inv *inv = (Mat_Inv*) imat->data;
+
+  PetscFunctionBegin;
+  if (!R) {
+    inv->R_id = -1;
+    inv->R_state = -1;
+    PetscFunctionReturn(0);
+  }
+  TRY( PetscObjectGetId((PetscObject)R, &inv->R_id) );
+  TRY( PetscObjectStateGet((PetscObject)R, &inv->R_state) );
+  PetscFunctionReturn(0);
+}
+
+//TODO we can take R from inv-A
+static PetscErrorCode MatInvNullSpaceSame_Inv(Mat imat, Mat R, PetscBool *flg)
```

Comment:
What are these two functions for? I.e., why does MatInv need to keep these state variables?

Reply:
If R changes, the regularization should be recomputed.

We could also lock further use of SetNullSpaceMat but that wouldn't solve the case when R is the same object but has changed - maybe it's fine?

Reply:
If R changes the K is still implicitly/explicitly regularized. The actual values in R do not matter as long as it represents a null space basis.

Reply:
OK, that makes sense to me. So I will remove this clutter.

Reply:
Done in 30c209e